examples/software-device-advanced/active_obj.cpp

```diff
@@ -0,0 +1,309 @@
+            if (bool(cur_options[rs::option::color_enable_auto_white_balance]))
+                sensors.at(RS2_STREAM_COLOR).set_option(RS2_OPTION_WHITE_BALANCE, legacy_dev->get_option(rs::option::color_white_balance));
+            if (bool(cur_options[rs::option::r200_lr_auto_exposure_enabled]))
+                sensors.at(RS2_STREAM_DEPTH).set_option(RS2_OPTION_EXPOSURE, legacy_dev->get_option(rs::option::r200_lr_exposure));
+            if (bool(cur_options[rs::option::fisheye_color_auto_exposure]))
+                sensors.at(RS2_STREAM_FISHEYE).set_option(RS2_OPTION_EXPOSURE, legacy_dev->get_option(rs::option::fisheye_exposure));
+        }
+        catch (...) {
+            // TODO: Handle errors somehow?
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(50));
+    }
+}
+
+legacy_active_obj::legacy_active_obj(rs::context& legacy_ctx, int legacy_dev_idx, rs2::software_device dev)
+    try : alive(true), legacy_dev(legacy_ctx.get_device(legacy_dev_idx))
+{
+    map_infos(dev);
+    map_profiles(dev);
+    map_options();
+} catch (...) {
+
+}
+
+void legacy_active_obj::finalize_binding(rs2::software_device dev) try {
+    // Set up destruction of active object when wrapper device is destroyed
+    auto shared_this = shared_from_this();
+    dev.set_destruction_callback([shared_this]() mutable { shared_this.reset(); });
+
+    heart = std::thread([this]() { heartbeat(); });
+} catch (std::exception &e) {
+    rs2::log(RS2_LOG_SEVERITY_ERROR, "Failed to register destruction callback to clean up legacy device's active object!");
+}
+
+void legacy_active_obj::map_infos(rs2::software_device dev) {
+    // Query legacy API for device infos and register them in wrapper
+    for (auto info : { rs::camera_info::serial_number, rs::camera_info::camera_type, rs::camera_info::oem_id }) {
+        if (!legacy_dev->supports(info)) continue;
+        dev.register_info(info_conv(info), legacy_dev->get_info(info));
+    }
+    // Set firmware version as both current and recommended versions
+    if (legacy_dev->supports(rs::camera_info::camera_firmware_version)) {
+        dev.register_info(RS2_CAMERA_INFO_FIRMWARE_VERSION, legacy_dev->get_info(rs::camera_info::camera_firmware_version));
+        dev.register_info(RS2_CAMERA_INFO_RECOMMENDED_FIRMWARE_VERSION, legacy_dev->get_info(rs::camera_info::camera_firmware_version));
+    }
+    // Device name is automatically initialized to "Software Device", so we update to overwrite that.
+    if (legacy_dev->supports(rs::camera_info::device_name)) {
+        dev.update_info(RS2_CAMERA_INFO_NAME, std::string("[Legacy] ") + legacy_dev->get_info(rs::camera_info::device_name));
+    }
+    // Set a couple other infos just for completion's sake
+    dev.register_info(RS2_CAMERA_INFO_ADVANCED_MODE, "NO");
+    dev.register_info(RS2_CAMERA_INFO_CAMERA_LOCKED, "NO");
+}
+
+void legacy_active_obj::map_profiles(rs2::software_device dev) {
```

Comment:
You modify a copy of sw_dev. is it by design?

Reply:
I'm not sure what you mean? rs2::software_device is a handle. I could change the function to pass by reference, but the internal logic isn't impacted by this